src/lib/tpm.c

```diff
@@ -29,339 +30,208 @@
-        .hierarchy = TPM2_RH_OWNER \
-    }, \
+    return rc;
 }
 
-#define SUPPORTED_ABI_VERSION \
-{ \
-    .tssCreator = 1, \
-    .tssFamily = 2, \
-    .tssLevel = 1, \
-    .tssVersion = 108, \
+TSS2_RC auth_cb (FAPI_CONTEXT *context, char const *description,
+                       char **auth, void *userData) {
+    (void)(context);
+    (void)(description);
+    *auth = userData;
+    return TSS2_RC_SUCCESS;
 }
 
-static ESYS_CONTEXT* esys_ctx_init(TSS2_TCTI_CONTEXT *tcti_ctx) {
-
-    TSS2_ABI_VERSION abi_version = SUPPORTED_ABI_VERSION;
-
-    ESYS_CONTEXT *esys_ctx = NULL;
-    TSS2_RC rval = Esys_Initialize(&esys_ctx, tcti_ctx, &abi_version);
-    if (rval != TPM2_RC_SUCCESS) {
-        LOGE("Esys_Initialize: 0x%x", rval);
-        return NULL;
-    }
-
-    return esys_ctx;
-}
+CK_RV tss_get_card_ids(CK_SLOT_ID *slot_list, CK_ULONG_PTR count) {
+    TSS2_RC rc;
+    FAPI_CONTEXT *fctx;
+    char *pathlist, *path, *strtokr_save = NULL;
+    size_t numPaths = 0;
+    CK_SLOT_ID tmp;
 
-static void flags_turndown(tpm_ctx *ctx, TPMA_SESSION flags) {
+    LOGV("Calling TSS to retrieve tokens.");
+    /* Let's check if everything is up and working.*/
+    rc = Fapi_Initialize(&fctx, NULL);
+    check_tssrc(rc, return CKR_GENERAL_ERROR);
 
-    if (ctx->esapi_manage_session_flags) {
-        return;
+    rc = Fapi_List(fctx, "/HS/SRK", &pathlist);
+    /* We immediately close FAPI because we have nowhere to put the context,
+       no global variables desired */
+    Fapi_Finalize(&fctx);
+    if (rc == 0x0006000a) {
+        numPaths = 0;
+        rc = TSS2_RC_SUCCESS;
     }
+    check_tssrc(rc, return CKR_GENERAL_ERROR);
 
-    TSS2_RC rc = Esys_TRSess_GetAttributes(ctx->esys_ctx, ctx->hmac_session, &ctx->old_flags);
-    assert(rc == TPM2_RC_SUCCESS);
-    if (rc != TSS2_RC_SUCCESS) {
-        LOGW("Esys_TRSess_SetAttributes: 0x%x", rc);
-        return;
-    }
+    LOGV("Processing TSS token results. keystore items found: %s", pathlist);
+    while ((path = strtok_r( (strtokr_save == NULL)? pathlist : NULL, ":", &strtokr_save))) {
+        if (sscanf(path, PREFIX "so-%08lx", &tmp) != 1) {
+            LOGV("%s is not a token, ignoring", path);
+            continue;
+        }
 
-    assert(ctx->old_flags == ctx->original_flags);
+        numPaths += 1;
+        LOGV("%s is token number %zi", path, numPaths);
 
-    TPMA_SESSION new_flags = (ctx->old_flags & (~flags));
-    rc = Esys_TRSess_SetAttributes(ctx->esys_ctx, ctx->hmac_session, new_flags, 0xff);
-    assert(rc == TSS2_RC_SUCCESS);
-    if (rc != TSS2_RC_SUCCESS) {
-        LOGW("Esys_TRSess_SetAttributes: 0x%x", rc);
-    }
-}
+        if (!slot_list)
+            continue;
 
-static void flags_restore(tpm_ctx *ctx) {
+        if (numPaths > *count) {
+            LOGE("buffer too small, count = %u, required=%u", *count, numPaths);
+            Fapi_Free(pathlist);
+            return CKR_BUFFER_TOO_SMALL;
+        }
```

Comment:
This has the same bug as https://github.com/tpm2-software/tpm2-pkcs11/issues/299

Reply:
fixed